File: python/2015_23.py
from aocd import get_data

inp = get_data(day=23, year=2015)

import re

instructions = [re.findall(r'[^, ]+', line) for line in inp.split('\n')]

def run_instructions(instructions, registers):
  i = 0
  while i < len(instructions):
    instruction, *args = instructions[i]
    
    if instruction == 'hlf':
      registers[args[0]] = registers[args[0]] / 2
    elif instruction == 'tpl':
      registers[args[0]] = registers[args[0]] * 3
    elif instruction == 'inc':
      registers[args[0]] += 1
    elif instruction == 'jmp':
      i += int(args[0]) - 1
    elif instruction == 'jie':
      if registers[args[0]] % 2 == 0:
        i += int(args[1]) - 1
    elif instruction == 'jio':
      if registers[args[0]] == 1:
        i += int(args[1]) - 1
    
    i += 1
    

registers = {'a': 0, 'b': 0}
run_instructions(instructions, registers)
answer = registers['b']
answer

registers = {'a': 1, 'b': 0}
run_instructions(instructions, registers)
answer = registers['b']
answer

File: R/2015_23.R
library(tidyverse)



df <-
  tibble(line = read_lines(input)) %>%
  transmute(
    line_number = row_number(),
    instruction = str_extract(line, "^[a-z]+"),
    r = str_extract(line, "(?<= )[a-z]+"),
    offset = str_extract(line, "[+-]\\d+") %>% parse_integer()
  )

df

hlf <- function(state, r, offset) {
  state[[r]] <- as.integer(state[[r]] / 2)
  state$line <- state$line + 1
  state
}

tpl <- function(state, r, offset) {
  state[[r]] <- state[[r]] * 3
  state$line <- state$line + 1
  state
}

inc <- function(state, r, offset) {
  state[[r]] <- state[[r]] + 1
  state$line <- state$line + 1
  state
}

jmp <- function(state, r, offset) {
  state$line <- state$line + offset
  state
}

jie <- function(state, r, offset) {
  if (state[[r]] %% 2 == 0) {
    state$line <- state$line + offset
  } else {
    state$line <- state$line + 1
  }
  state
}

jio <- function(state, r, offset) {
  if (state[[r]] == 1) {
    state$line <- state$line + offset
  } else {
    state$line <- state$line + 1
  }
  state
}

state <- list(line = 1, a = 0, b = 0)

while (state$line <= nrow(df)) {
  inst <- df %>% slice(state$line)
  
  f <- get(inst$instruction)
  state <- f(state, inst$r, inst$offset)
}

answer <- state$b
answer

state <- list(line = 1, a = 1, b = 0)

while (state$line <= nrow(df)) {
  inst <- df %>% slice(state$line)
  
  f <- get(inst$instruction)
  state <- f(state, inst$r, inst$offset)
}

answer <- state$b
answer
